README.pod
=pod

=encoding UTF-8

=for :stopwords Apocalypse Niebur Ryan cpan testmatrix url annocpan anno bugtracker rt
cpants kwalitee diff irc mailto metadata placeholders metacpan apocal CPAN
AUTHORs al backend debian distro distros dists env hackish plugins
testsuite yml pm yay unicode blog precompiled =for stopwords ap cyclomatic
cal dist homebrew imo internet perltidy prefs prereq testrun

=head1 NAME

Test::Apocalypse - Apocalypse's favorite tests bundled into a simple interface

=head1 VERSION

  This document describes v1.006 of Test::Apocalypse - released October 25, 2014 as part of Test-Apocalypse.

=head1 SYNOPSIS

	#!/usr/bin/perl
	use strict; use warnings;

	use Test::More;
	eval "use Test::Apocalypse";
	if ( $@ ) {
		plan skip_all => 'Test::Apocalypse required for validating the distribution';
	} else {
		is_apocalypse_here();
	}

=head1 DESCRIPTION

This module greatly simplifies common author tests for modules heading towards CPAN. I was sick of copy/pasting
the tons of t/foo.t scripts + managing them in every distro. I thought it would be nice to bundle all of it into
one module and toss it on CPAN :) That way, every time I update this module all of my dists would be magically
updated!

This module respects the RELEASE_TESTING/AUTOMATED_TESTING env variable, if it is not set it will skip the entire
testsuite. Normally end-users should not run it; but you can if you want to see how bad my dists are, ha! The scheme
is exactly the same as the one Alias proposed in L<Test::XT> and in his blog post, L<http://use.perl.org/use.perl.org/_Alias/journal/38822.html>.

This module uses L<Module::Pluggable> to have custom "backends" that process various tests. We wrap them in a hackish
L<Test::Block> block per-plugin and it seems to work nicely. If you want to write your own, it should be a breeze
once you look at some of my plugins and see how it works. ( more documentation to come )

=head2 Usage

In order to use this, you would need to be familiar with the "standard" steps in order to fully exercise the testsuite.
There are a few steps we require, because our plugins need stuff to be prepared for them. For starters, you would need
a test file in your distribution similar to the one in SYNOPSIS. Once that is done and added to your MANIFEST and etc,
you can do this:

	perl Build.PL			# sets up the dist ( duh, hah )
	./Build dist			# makes the tarball ( so certain plugins can process it )
	RELEASE_TESTING=1 ./Build test	# runs the testsuite!

=head1 Methods

=head2 is_apocalypse_here()

This is the main entry point for this testsuite. By default, it runs every plugin in the testsuite. You can enable/disable
specific plugins if you desire. It accepts a single argument: a hashref or a hash. It can contain various options, but as of
now it only supports two options. If you try to use allow and deny at the same time, this module will throw an exception.

=head3 allow

Setting "allow" to a string or a precompiled regex will run only the plugins that match the regex. If passed a string, this module
will compile it via C<qr/$str/i>.

	# run only the EOL test and disable all other tests
	is_apocalypse_here( {
		allow	=> qr/^EOL$/,
	} );

	# run all "dist" tests
	is_apocalypse_here( {
		allow	=> 'dist',
	} );

=head3 deny

Setting "deny" to a string or a precompiled regex will not run the plugins that match the regex. If passed a string, this module
will compile it via C<qr/$str/i>.

	# disable Pod_Coverage test and enable all other tests
	is_apocalypse_here( {
		deny	=> qr/^Pod_Coverage$/,
	} );

	# disable all pod tests
	is_apocalypse_here( {
		deny	=> 'pod',
	} );

=head2 plugins()

Since this module uses L<Module::Pluggable> you can use this method on the package to find out what plugins are available. Handy if you need
to know what plugins to skip, for example.

	my @tests = Test::Apocalypse->plugins;

=head1 EXPORT

Automatically exports the "is_apocalypse_here" sub.

=head1 MORE IDEAS

=over 4

=item * Test::NoSmartComments

I don't use Smart::Comments but it might be useful? I LOVE BLOAT! :)

=item * Better POD spelling checker?

Test::Spelling is ancient, and often blows up. There's a Test::Pod::Spelling on CPAN but it is flaky too :(

=item * Document the way we do plugins so others can add to this testsuite :)

=item * POD standards check

Do we have SYNOPSIS, ABSTRACT, SUPPORT, etc sections? ( PerlCritic can do that! Need to investigate more... )

=item * Integrate Test::UniqueTestNames into the testsuite

This would be nice, but I'm not sure if I can actually force this on other tests. Otherwise I'll be just making
sure that the Test::Apocalypse tests is unique, which is worthless to $dist trying to clean itself up...

=item * META.yml checks

We should make sure that the META.yml includes the "repository", "license", and other useful keys!

=item * Other AUTHORs

As always, we should keep up on the "latest" in the perl world and look at other authors for what they are doing.

=item * indirect syntax

We should figure out how to use indirect.pm to detect this deprecated method of coding. There's a L<Perl::Critic> plugin for this, yay!

=item * Test::PPPort

Already implemented as PPPort.pm but it's less invasive than my version, ha!

=item * Test::DependentModules

This is a crazy test, but would help tremendously in finding regressions in your code!

=item * Test::CleanNamespaces

I don't exclusively code in Moose, but this could be useful...

=item * no internet?

It would be nice to signal INTERNET_TESTING=0 or something zany like that so this testsuite will skip the tests that need internet access...

	<Apocalypse> Is there a convention that signals no internet access? Similar to RELEASE_TESTING, AUTOMATED_TESTING, and etc?
	<@rjbs> No.
	<Apocalypse> mmm I ain't in the mood to invent it so I'll just bench it for now :(
	<Apocalypse> however, if I was to invent it I would call it something like INTERNET_TESTING=0
	<Apocalypse> Also, why does ILYAZ keep re-inventing the stuff? Use of uninitialized value $ENV{"PERL_RL_TEST_PROMPT_MINLEN"} in bitwise or (|) at test.pl line 33.
	<@Alias> use LWP::Online ':skip_all';
	<@Alias> Whack that in the relevant test scripts
	<Apocalypse> Alias: Hmm, how can I control that at a distance? i.e. disabling inet if I had inet access?
	<@Alias> You can't
	<@Alias> It's a pragmatic test, tries to pull some huge site front pages and looks for copyright statements
	<Apocalypse> At least it's a good start - thanks!
	<@Alias> So it deals with proxies and airport wireless hijacking etc properly
	<Apocalypse> Hah yeah I had to do the same thing at $work in the past, we put up a "special" page then had our software try to read it and if the content didn't match it complained :)
	<@Alias> right
	<@Alias> So yeah, it automates that
	<@Alias> I wrote it while in an airport annoyed that something I wrote wasn't falling back on a minicpan properly
	<Apocalypse> At least it'll be an improvement, but I still need to force no inet for testing... ohwell
	<Apocalypse> Heh, it seems like us perl hackers do a lot of work while stranded at airports :)
	<@Alias> If you can break LWP from the environment, that would work
	<@Alias> Setting a proxy ENVthat is illegal etc
	<Apocalypse> ah good thinking, I'll read up on the fine points of LWP env vars and try to screw it up

=back

=head2 Modules that I considered but decided against using

=over 4

=item * L<Test::Distribution>

This module was a plugin in this testsuite but I don't need it. All the functionality in it is already replicated in the plugins :)

=item * L<Test::Module::Used> and L<Test::Dependencies>

They were plugins in this testsuite but since I started coding with L<Moose>, they don't work! I've switched to my homebrew solution
utilizing L<Perl::PrereqScanner> which works nicely for me.

=item * L<Test::MyDeps>

Superseded by L<Test::DependentModules>. Also, I don't want to waste a lot of time on each testrun testing other modules!

=item * L<Test::NoTabs>

I always use tabs! :(

=item * L<Test::CheckManifest>

This was a buggy module that I dropped and is now using L<Test::DistManifest>

=item * L<Test::Dist>

This is pretty much the same thing as this dist ;)

=item * L<Test::PureASCII>

This rocks, as I don't care about unicode in my perl! ;)

=item * L<Test::LatestPrereqs>

This looks cool but we need to fiddle with config files? My OutdatedPrereqs test already covers it pretty well...

=item * L<Test::Pod::Content>

This is useful, but not everyone has the same POD layout. It would be too much work to try and generalize this...

=item * L<Test::GreaterVersion>

Since I never use CPAN, this is non-functional for me. However, it might be useful for someone?

=item * L<Test::Kwalitee>

This dist rocks, but it doesn't print the info nor utilize the extra metrics. My homebrew solution actually copied
a lot of code from this, so I have to give it props!

=item * L<Test::LoadAllModules>

This is very similar to L<Test::UseAllModules> but looks more complicated. Also, I already have enough tests that do that ;)

=item * L<Test::ModuleReady>

This looks like a nice module, but what it does is already covered by the numerous tests in this dist...

=item * L<Test::PerlTidy>

Br0ken install at this time... ( PerlCritic can do that! Need to investigate more... ) Also, all it does is... run your module
through perltidy and compare the outputs. Not that useful imo because I never could get perltidy to match my prefs :(

=item * L<Test::Install::METArequires>

This looks like a lazy way to do auto_install and potentially dangerous! Better to just use the prereq logic in Build.PL/Makefile.PL

=item * L<Test::Perl::Metrics::Simple>

This just tests your Cyclomatic complexity and was the starting point for my homebrew solution.

=back

=head1 SUPPORT

=head2 Perldoc

You can find documentation for this module with the perldoc command.

  perldoc Test::Apocalypse

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

MetaCPAN

A modern, open-source CPAN search engine, useful to view POD in HTML format.

L<http://metacpan.org/release/Test-Apocalypse>

=item *

Search CPAN

The default CPAN search engine, useful to view POD in HTML format.

L<http://search.cpan.org/dist/Test-Apocalypse>

=item *

RT: CPAN's Bug Tracker

The RT ( Request Tracker ) website is the default bug/issue tracking system for CPAN.

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-Apocalypse>

=item *

AnnoCPAN

The AnnoCPAN is a website that allows community annotations of Perl module documentation.

L<http://annocpan.org/dist/Test-Apocalypse>

=item *

CPAN Ratings

The CPAN Ratings is a website that allows community ratings and reviews of Perl modules.

L<http://cpanratings.perl.org/d/Test-Apocalypse>

=item *

CPAN Forum

The CPAN Forum is a web forum for discussing Perl modules.

L<http://cpanforum.com/dist/Test-Apocalypse>

=item *

CPANTS

The CPANTS is a website that analyzes the Kwalitee ( code metrics ) of a distribution.

L<http://cpants.perl.org/dist/overview/Test-Apocalypse>

=item *

CPAN Testers

The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

L<http://www.cpantesters.org/distro/T/Test-Apocalypse>

=item *

CPAN Testers Matrix

The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

L<http://matrix.cpantesters.org/?dist=Test-Apocalypse>

=item *

CPAN Testers Dependencies

The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

L<http://deps.cpantesters.org/?module=Test::Apocalypse>

=back

=head2 Email

You can email the author of this module at C<APOCAL at cpan.org> asking for help with any problems you have.

=head2 Internet Relay Chat

You can get live help by using IRC ( Internet Relay Chat ). If you don't know what IRC is,
please read this excellent guide: L<http://en.wikipedia.org/wiki/Internet_Relay_Chat>. Please
be courteous and patient when talking to us, as we might be busy or sleeping! You can join
those networks/channels and get help:

=over 4

=item *

irc.perl.org

You can connect to the server at 'irc.perl.org' and join this channel: #perl-help then talk to this person for help: Apocalypse.

=item *

irc.freenode.net

You can connect to the server at 'irc.freenode.net' and join this channel: #perl then talk to this person for help: Apocal.

=item *

irc.efnet.org

You can connect to the server at 'irc.efnet.org' and join this channel: #perl then talk to this person for help: Ap0cal.

=back

=head2 Bugs / Feature Requests

Please report any bugs or feature requests by email to C<bug-test-apocalypse at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Test-Apocalypse>. You will be automatically notified of any
progress on the request by the system.

=head2 Source Code

The code is open to the world, and available for you to hack on. Please feel free to browse it and play
with it, or whatever. If you want to contribute patches, please send me a diff or prod me to pull
from your repository :)

L<https://github.com/apocalypse/perl-test-apocalypse>

  git clone git://github.com/apocalypse/perl-test-apocalypse.git

=head1 AUTHOR

Apocalypse <APOCAL@cpan.org>

=head2 CONTRIBUTORS

=for stopwords Apocalypse Ryan Niebur

=over 4

=item *

Apocalypse <apoc@blackhole.(none)>

=item *

Apocalypse <apoc@satellite.(none)>

=item *

Apocalypse <perl@0ne.us>

=item *

Ryan Niebur <ryanryan52@gmail.com>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to jawnsy@cpan.org for the prodding and help in getting this package ready to be bundled into debian!

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Apocalypse.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

The full text of the license can be found in the
F<LICENSE> file included with this distribution.

=head1 DISCLAIMER OF WARRANTY

THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT
HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY
OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM
IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF
ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS
THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY
GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE
USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF
DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD
PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS),
EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

=cut


Changes
Revision history for Test::Apocalypse

{{$NEXT}}

	Removed yet another perlcritic rule, thanks to crazy UTF8 creeping into my code :)
	Fixed the Dependencies plugin so it also processes the bin/ directory
	Fixed the Dependencies plugin to skip testing Task/Bundle dists

1.006	2014-10-25 20:10:12 UTC

	Proper release on CPAN :)

1.005	2014-10-25 19:58:22 UTC

	FakeRelease to test my dzil configs!

1.004	2014-10-25 05:07:26 UTC

	Tweak how we skip/disable plugins because of the recent subtest change, I broke some stuff :(
	I give up on the Pod::No404s test, the internet is too flaky everywhere :(
	Removed the OutdatedPrereqs test because I now use Dist::Zilla::Plugin::PromptIfStale, thanks ETHER!

1.003	2014-10-25 00:12:08 UTC

	Add Test::Mojibake to the dist
	Add more exclusions to the standard Perl::Critic config
	Add PERL_CRITIC_STRICT env var to skip the bundled Perl::Critic config
	Fix an error in Test::Apocalypse::Script
	Use Test::CPAN::Changes instead of the outdated Test::CheckChanges
	Ripped out the dysfunctional Test::NoWarnings in favor of Test::FailWarnings, thanks DAGOLDEN!
	Ignore the 'perl' dependency when checking deps
	Fix version mismatch error when checking dependencies ( 'v0.0.6' vs '0.0.6' )
	Thanks everyone who sent RT tickets to clean up the FAILs :)
	Finally enabled the Test::Vars plugin as I've switched over to using subtests instead of Test::Block

1.002	2011-04-21 21:49:46 UTC

	Add more default exclusions to the Perl::Critic default .perlcriticrc
	Tweaked the OutdatedPrereqs test *again* so it FAILs if a prereq version is higher than found on CPAN
	Tweaked the Dependencies test so it correctly processes found prereqs that has no version

1.001	2011-03-08 21:37:20 UTC

	Tweaked the OutdatedPrereqs test so it will correctly FAIL if a prereq wasn't found on CPAN
	Tweaked the Dependencies test so it correctly filters provided packages

1.000	2011-03-04 23:37:23 UTC

	Converted to Dist::Zilla for the release process!
	Tweaked the internals to cleanup automated issues
	Ported the META.(yml|json) tests to the new Test::CPAN::Meta infrastructure, thanks JAWNSY! ( RT#66215 )
	Changed some plugins so they execute tests as TODO, to not cause spurious FAILs
	Added Test::Vars to the testsuite
	Added Test::Script to the testsuite
	Added Test::Pod::LinkCheck to the testsuite
	Added Test::Pod::Spelling::CommonMistakes to the testsuite
	Added debugging of test info, thanks POE::Test::Loops!
	Replaced Test::Dependencies with homebrew solution utilizing Perl::PrereqScanner
	Removed Test::Module::Used plugin ( see above )
	Removed the Test::Prereq plugin because it never worked reliably and we already have a plugin for it ( see above )
	Removed the Test::UseAllModules plugin because we have other plugins that do the same thing
	Bumped Test::NoPlan prereq to 0.0.6 so it works properly under Test::Block
	Bumped Test::Pod prereq to 1.41 so it doesn't complain about L<text|scheme> being invalid
	Added Pod::Coverage::TrustPod as the coverage class for Test::Pod::Coverage
	Fixed a bug in Test::Spelling where the file stopwords weren't added right
	Fixed a bug in OutdatedPrereqs where it complained if the CPAN version was older than what was specified
	Revamped the Perl::Critic test to automatically use a default config if one wasn't found in t/perlcriticrc

0.10	0000-00-00 00:00:00 UTC

	Added smart AUTOMATED_TESTING logic so we are now behaving in the same way as Alias envisaged
		This means certain tests are disabled when AUTOMATED is set :)
	Minor POD/Build.PL/test output tweaks
	Added META.json to FileChecks test
	Bumped Test::Strict prereq to 0.14 because now it works with Moose!
	Tweaked the module loading logic so we have less duplication now + less bugs, yay!
	Fixed a warning in the OutdatedPrereqs test so we don't cause FAILs with the Test::NoWarnings module :(
	Added Test::NoPlan to the testsuite

0.09	0000-00-00 00:00:00 UTC

	Added the PERL_APOCALYPSE env var to override test selection and test all plugins
	Run the plugins in alphabetical order, which is prettier :)
	Removed the crappy stopwords addition in Pod_Spelling test
	Added Test::JSON::Meta to the testsuite

0.08	0000-00-00 00:00:00 UTC

	Minor POD tweaks
	Added Test::Pod::No404s to the testsuite
	Fixed a warning in the PerlMetrics test so we don't cause FAILs with the Test::NoWarnings module :(

0.07	0000-00-00 00:00:00 UTC

	Fixed a issue with META.yml quoting causing parse failures when using tools that depend on it
	Upgraded Module::Build to 0.3601, had to change prereq from '0.2.2' to '0.002002' to satisfy M::B::Compat :(

0.06	0000-00-00 00:00:00 UTC

	Fixed OutdatedPrereqs test so it doesn't blow up on zero prereqs - thanks Ryan52!
	Added $@ to the module load failure output, for better debugging - thanks Ryan52!
	Minor POD tweaks
	Added more common terms to the Pod_Spelling test
	Added Test::NoBreakpoints to the testsuite
	Added Test::ConsistentVersion to the testsuite
	Added Test::EOL to the testsuite
	Added Test::AutoLoader to the testsuite
	Modified is_apocalypse_here() to accept options, you can now enable/disable specific tests!
	Test::Apocalypse now finally passes it's own testsuite under RELEASE_TESTING, ha! ( thanks to Debian guys for prodding me )
	The Kwalitee test for EXPERIMENTAL/EXTRA metrics is not treated as a failure now
	Oops, Module::Pluggable wasn't listed as a prereq!
	Sorry - added a lot of Perl::Critic extensions so you will not get failed tests because you missed to test an extension!

0.05	0000-00-00 00:00:00 UTC

	Added Test::Module::Used to the testsuite
	Added Test::Synopsis to the testsuite
	Enabled the Test::Pod::Coverage test, it was previously disabled by mistake!
	Changed to the new ENV var scheme, proposed by Alias ( and thanks to JAWNSY for prodding me )

0.04	0000-00-00 00:00:00 UTC

	Removed Test::CheckManifest and replaced with Test::DistManifest, thanks JAWNSY!
	Various POD cleanups and additions

0.03	0000-00-00 00:00:00 UTC

	Added more tests :)


lib/Test/Apocalypse/EOL.pm
package Test::Apocalypse::EOL;

# ABSTRACT: Plugin for Test::EOL

use Test::EOL 0.3;

# TODO If a win32 user downloads the tarball, it will have DOS newlines in it?
sub _do_automated { 0 }

sub do_test {
	all_perl_files_ok();

	return;
}

1;

=pod

=for Pod::Coverage do_test

=head1 DESCRIPTION

Encapsulates L<Test::EOL> functionality.

=cut


lib/Test/Apocalypse/Pod.pm
package Test::Apocalypse::Pod;

# ABSTRACT: Plugin for Test::Pod

use Test::Pod 1.41;

sub do_test {
	all_pod_files_ok();

	return;
}

1;

=pod

=for Pod::Coverage do_test

=head1 DESCRIPTION

Encapsulates L<Test::Pod> functionality.

=cut


Build.PL
# Build.PL
use strict; use warnings;
use Module::Build;

my $build = Module::Build->new(
	# look up Module::Build::API for the info!
	'dynamic_config'	=> 0,
	'module_name'		=> 'Test::Apocalypse',
	'license'		=> 'perl',

	'dist_abstract'		=> "Apocalypse's favorite tests bundled into a simple interface",
	'dist_author'		=> 'Apocalypse <APOCAL@cpan.org>',

	'create_packlist'	=> 1,
	'create_makefile_pl'	=> 'traditional',
	'create_readme'		=> 1,
	'create_license'	=> 1,
	'sign'			=> 0,

	'test_files'		=> 't/*.t',

	'requires'		=> {
		# FIXME this sucks, I have to repeat the version here and in the Test::Apocalypse::XYZ module...
		# Dist::Zilla will solve it! I need to switch someday :)

		# Test stuff
		'Test::More'			=> '0.94',
		'Test::Compile'			=> '0.11',
		'Test::Perl::Critic'		=> '1.02',
		'Test::Distribution'		=> '2.00',
		'Test::Fixme'			=> '0.04',
		'Test::HasVersion'		=> '0.012',
		'Test::MinimumVersion'		=> '0.101080',
		'Test::Pod::Coverage'		=> '1.08',
		'Test::Spelling'		=> '0.11',
		'Test::Pod'			=> '1.41',
		'Test::Prereq'			=> '1.037',
		'Test::Prereq::Build'		=> '1.037',
		'Test::Strict'			=> '0.14',
		'Test::UseAllModules'		=> '0.12',
		'Test::YAML::Meta'		=> '0.14',
		'Test::NoWarnings'		=> '0.084',
		'Test::CheckChanges'		=> '0.08',
		'Test::File'			=> '1.29',
		'Test::Dir'			=> '1.006',
		'Test::Signature'		=> '1.10',
		'Test::Portability::Files'	=> '0.05',
		'Test::DistManifest'		=> '1.005',
		'Test::Synopsis'		=> '0.06',
		'Test::CPAN::Meta'		=> '0.13',
		'Test::NoBreakpoints'		=> '0.13',
		'Test::EOL'			=> '0.3',
#		'Test::ConsistentVersion'	=> '0.2.2',	# M::B::Compat 0.36 doesn't like this!
		'Test::ConsistentVersion'	=> '0.002002',
		'Test::AutoLoader'		=> '0.03',
		'Test::Pod::No404s'		=> '0.01',
		'Test::JSON::Meta'		=> '0.04',
		'Test::NoPlan'			=> '0.000006',	# Same M::B::Compat issue
		'Test::Vars'			=> '0.001',
		'Test::Script'			=> '1.07',
		'Pod::Coverage::TrustPod'	=> '0.092830',
		'Test::Pod::LinkCheck'		=> '0.004',
		'Test::Deep'			=> '0.108',

		# handy tools
		'File::Spec'			=> '3.31',
		'File::Find::Rule'		=> '0.32',
		'Module::CPANTS::Analyse'	=> '0.85',
		'Perl::Metrics::Simple'		=> '0.13',
		'Test::Block'			=> '0.11',
		'YAML'				=> '0.70',
		'version'			=> '0.77',
		'Devel::PPPort'			=> '3.19',
		'Module::CoreList'		=> '2.23',
		'Module::Pluggable'		=> '3.9',
		'File::Which'			=> '1.09',
		'File::Slurp'			=> '9999.13',
		'YAML::Any'			=> '0.72',
		'JSON::Any'			=> '1.25',
		'Perl::PrereqScanner'		=> '1.000',

		# We can't specify the CPANPLUS::Configure/Backend module because...
#apoc@blackhole:~/mygit/perl-test-apocalypse$ perl Build.PL
#Checking prerequisites...
#  requires:
#    !  CPANPLUS::Backend (0) is installed, but we need version >= 0.90
#    !  CPANPLUS::Configure (0) is installed, but we need version >= 0.90
		'CPANPLUS'			=> '0.90',

		# Perl::Critic extensions so we're sure we are testing every one of them, ha!
		'Task::Perl::Critic'		=> '1.007',

		# we need a recent perl
		'perl'				=> '5.008',
	},

	# include the standard stuff in META.yml
	'meta_merge'	=> {
		'resources'	=> {
			'license'	=> 'http://dev.perl.org/licenses/',
			'homepage'	=> 'http://search.cpan.org/dist/Test-Apocalypse',
			'bugtracker'	=> 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-Apocalypse',
			'repository'	=> 'http://github.com/apocalypse/perl-test-apocalypse',
		},
	},
);

# all done!
$build->create_build_script;
